tests/test_love_manager_client.py

```diff
@@ -0,0 +1,316 @@
+        }
+
+    def purge_received_data(self):
+        self.received_data = dict()
+
+    async def wait_for_number_of_samples(self, number_of_samples, sample_type="event"):
+
+        self.log.debug(
+            f"Waiting for {number_of_samples} samples or {self.standard_timeout} seconds, "
+            "whatever comes first."
+        )
+
+        test_timeout_task = asyncio.create_task(asyncio.sleep(self.standard_timeout))
+
+        while (not test_timeout_task.done()) and (
+            len(self.received_data.get(sample_type, [])) < number_of_samples
+        ):
+            await asyncio.sleep(self.pool_timeout)
+
+        await cancel_task(test_timeout_task)
+
+        self.log.debug(
+            f"Done waiting for {number_of_samples} samples. "
+            f"Got {len(self.received_data.get(sample_type, []))}."
+        )
+
+    @contextlib.asynccontextmanager
+    async def setup_test_environment_to_handle_connection(self):
+        async with self.configure_environment_with_websocket_host():
+            async with self.start_mock_manager() as socket:
+                async with self.handle_connection_with_manager():
+                    yield socket
+
+    @contextlib.asynccontextmanager
+    async def configure_environment_with_websocket_host(self):
+
+        back_up_websocket_host = (
+            os.environ.pop("WEBSOCKET_HOST") if "WEBSOCKET_HOST" in os.environ else None
+        )
+
+        os.environ["WEBSOCKET_HOST"] = "0.0.0.0:9999"
+
+        try:
+            yield
+        finally:
+            if back_up_websocket_host is not None:
+                os.environ["WEBSOCKET_HOST"] = back_up_websocket_host
+            else:
+                os.environ.pop("WEBSOCKET_HOST")
+
+    @contextlib.asynccontextmanager
+    async def start_mock_manager(self):
+
+        self._run_wesocket_server = True
+
+        socket = websockets.serve(
+            ws_handler=self.handle_websocket_server_received_message,
+            host="0.0.0.0",
+            port=9999,
+        )
+
+        try:
+            await socket
+            yield socket
+        finally:
+            self.log.info("Closing web server")
+            self._run_wesocket_server = False
+            socket.ws_server.close()
+
+    @contextlib.asynccontextmanager
+    async def handle_connection_with_manager(self):
+
+        try:
+            start_task = asyncio.create_task(
+                self.love_manager_client.handle_connection_with_manager()
+            )
+
+            while self.love_manager_client.connected_task is None:
+
+                await asyncio.sleep(self.pool_timeout)
+
+            await asyncio.wait_for(
+                self.love_manager_client.connected_task,
+                timeout=self.standard_timeout,
+            )
+
+            yield
+
+        finally:
+            await self.love_manager_client.close()
+            await asyncio.wait_for(start_task, timeout=self.standard_timeout)
```

Comment:
How does this work?

Reply:
So, `start_task` has the task returned by:

```
start_task = asyncio.create_task(
                self.love_manager_client.handle_connection_with_manager()
            )
```

above. 

This task will execute until we call `self.love_manager_client.close()`, which is done on the line above. 

When we call `await self.love_manager_client.close()`, the class will start shutting down and cleaning up.

If something wrong happened and it class is stuck cleaning up, I want to interrupt it. The `await_for` does exactly that. If `start_task` does no finish in time it will be cancelled and I will get a timeout error.

Reply:
I see... thanks for the clear explanation!